presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/PinotClusterInfoFetcher.java

```diff
@@ -486,4 +496,101 @@ public TimeBoundary getTimeBoundaryForTable(String table)
             throw e;
         }
     }
+
+    public static class Instance
+    {
+        private final String instanceName;
+        private final String hostName;
+        private final boolean enabled;
+        private final int port;
+        private final int grpcPort;
+        private final List<String> tags;
+        private final List<String> pools;
+
+        @JsonCreator
+        public Instance(
+                @JsonProperty String instanceName,
+                @JsonProperty String hostName,
+                @JsonProperty boolean enabled,
+                @JsonProperty int port,
+                @JsonProperty int grpcPort,
+                @JsonProperty List<String> tags,
+                @JsonProperty List<String> pools)
+        {
+            this.instanceName = instanceName;
+            this.hostName = hostName;
+            this.enabled = enabled;
+            this.port = port;
+            this.grpcPort = grpcPort;
+            this.tags = tags;
+            this.pools = pools;
+        }
+
+        public String getInstanceName()
+        {
+            return instanceName;
+        }
+
+        public String getHostName()
+        {
+            return hostName;
+        }
+
+        public boolean isEnabled()
+        {
+            return enabled;
+        }
+
+        public int getPort()
+        {
+            return port;
+        }
+
+        public int getGrpcPort()
+        {
+            return grpcPort;
+        }
+
+        public List<String> getTags()
+        {
+            return tags;
+        }
+
+        public List<String> getPools()
+        {
+            return pools;
+        }
+    }
+
+    public Instance getInstance(String instanceName)
+    {
+        try {
+            String responseBody = sendHttpGetToController(String.format(INSTANCE_API_TEMPLATE, instanceName));
+            return instanceJsonCodec.fromJson(responseBody);
+        }
+        catch (Exception throwable) {
+            throw new PinotException(PINOT_UNABLE_TO_FIND_INSTANCE, Optional.empty(), "Error when fetching instance configs for " + instanceName, throwable);
+        }
+    }
+
+    public int getGrpcPort(String serverInstance)
+    {
+        try {
+            int grpcPort = instanceConfigCache.get(serverInstance).getGrpcPort();
```

Comment:
Can you push this logic of choosing the global rpc port to when the instance config cache is populated ? Either the controller provides you the rpc port, or you choose the global one. I feel it belongs better at "population" time than at "read time".

Reply:
make sense, moved this logic into `PinotConnection.getGrpcPort(String serverInstance)` to apply the pinot configs.
For `PinotClusterInfoFetcher`, we just honor the real value fetched from Pinot cluster.